src/notation/view/pianokeyboard/pianokeyboardview.cpp

```diff
@@ -0,0 +1,476 @@
+
+            bottomPieceGradient.setColorAt(1.0, backgroundColor);
+            bottomPieceGradient.setStart(0.0, top);
+            bottomPieceGradient.setFinalStop(0.0, bottom);
+        }
+
+        QColor highlightColor = m_blackKeyStateColors[m_controller->keyState(key)];
+        topPieceGradient.setColorAt(1.0, highlightColor);
+        bottomPieceGradient.setColorAt(0.0, highlightColor);
+
+        painter->translate(rect.topLeft());
+        painter->fillRect(backgroundRect, backgroundColor);
+        painter->fillPath(topPiecePath, topPieceGradient);
+        painter->fillPath(bottomPiecePath, bottomPieceGradient);
+        painter->translate(-rect.topLeft());
+    }
+}
+
+int PianoKeyboardView::numberOfKeys() const
+{
+    return m_numberOfKeys;
+}
+
+void PianoKeyboardView::setNumberOfKeys(int number)
+{
+    if (m_numberOfKeys == number) {
+        return;
+    }
+
+    switch (number) {
+    case 128:
+        m_lowestKey = 0;
+        m_numberOfKeys = 128;
+        break;
+    case 88:
+        m_lowestKey = 21;
+        m_numberOfKeys = 88;
+        break;
+    case 61:
+        m_lowestKey = 36;
+        m_numberOfKeys = 61;
+        break;
+    case 49:
+        m_lowestKey = 36;
+        m_numberOfKeys = 49;
+        break;
+    case 25:
+        m_lowestKey = 48;
+        m_numberOfKeys = 25;
+        break;
+    default:
+        return;
+    }
+
+    emit numberOfKeysChanged();
+    calculateKeyRects();
+    update();
+}
+
+void PianoKeyboardView::moveCanvas(qreal dx)
+{
+    m_scrollOffset += dx;
+    adjustKeysAreaPosition();
+    update();
+}
+
+qreal PianoKeyboardView::keyWidthScaling() const
+{
+    return m_keyWidthScaling;
+}
+
+void PianoKeyboardView::scale(qreal factor, qreal x)
+{
+    setScaling(m_keyWidthScaling * factor, x);
+}
+
+void PianoKeyboardView::setScaling(qreal scaling, qreal x)
+{
+    qreal newScaling = std::clamp(scaling, 0.5, 2.0);
+    qreal correctedFactor = newScaling / m_keyWidthScaling;
+
+    m_keyWidthScaling = newScaling;
+    m_scrollOffset *= correctedFactor;
+    m_scrollOffset += x * (1 - correctedFactor);
+
+    emit keyWidthScalingChanged();
+    calculateKeyRects();
+    update();
+}
+
+qreal abs2d(qreal x, qreal y)
+{
+    return sqrt(x * x + y * y) * (y > -x ? 1 : -1);
+}
+
+qreal PianoKeyboardView::scrollBarPosition() const
+{
+    return m_scrollBarPosition;
+}
+
+qreal PianoKeyboardView::scrollBarSize() const
+{
+    return m_scrollBarSize;
+}
+
+void PianoKeyboardView::updateScrollBar()
+{
+    qreal newPosition = -m_scrollOffset / m_keysAreaRect.width();
+    qreal newSize = width() / m_keysAreaRect.width();
+
+    if (qFuzzyCompare(newPosition, m_scrollBarPosition)
+        && qFuzzyCompare(newSize, m_scrollBarSize)) {
+        return;
+    }
+
+    m_scrollBarPosition = newPosition;
+    m_scrollBarSize = newSize;
+    emit scrollBarChanged();
+}
+
+void PianoKeyboardView::setScrollBarPosition(qreal position)
+{
+    m_scrollOffset = -position* m_keysAreaRect.width();
```

Comment:
`-position* m_keysAreaRect.width();`

Wrong indentation :)

Also, it is better to check if scollOffset was actually changed before calling update()

```
qreal newScrollOffset = -position * m_keysAreaRect.width();
if (m_scrollOffset == newScrollOffset) {
    return;
}

adjustKeysAreaPosition();
update();
```